### PR TITLE
frontend: check liquidity on target chain

### LINKED ISSUE #2141: frontend: disable transfer creation if not enough liquidity available
## Issue Desription

We would like to disable the transfer submit button in case the transfer cannot get filled due to liquidity. We should check the token balance of all agents on the target chain and disable if none of them has enough tokens for fulfillment. For this a new validator function needs to be implemented and integrated into the form.

## Acceptance Criteria
- [ ] check the token balances of the agents for the selected token on the target chain
- [ ] disable transfer button in case no agent has as much tokens as the user's input in the amount field

### PR DESCRIPTION
**Pull request by manuelwedler**
closes #2141 

Disables transfer button if not enough liquidity on target chain. Checks against the liquidity provider with the highest amount of tokens. Agent addresses are hardcoded. 

I started to refactor our provider classes a little in the way it was discussed earlier. It was much easier to implement this functionality when having an option to create providers from an rpc url. This allows now to get rid of many of the JsonRpcProvider objects. Ideally we would have a central place (a store) with one provider per chain which use our internal rpc urls.

### CODE REVIEW

File: frontend/src/composables/useMaxFillableAmount.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  const provider: Ref<IEthereumProvider | undefined> = shallowRef(undefined);
+  watch(
+    targetChain,
+    async (chain) =>
+      (provider.value = chain ? await createEthereumProvider(chain.internalRpcUrl) : undefined),
+  );
+
+  const balances: Ref<TokenAmount | undefined>[] = [];
+  for (const agent of AGENT_ADDRESSES) {
+    const { balance } = useTokenBalance(provider, ref(agent), token);
```
Comment (compojoom) [inline frontend/src/composables/useMaxFillableAmount.ts:24]: Can you explain how this work? To me it is very confusing. The second argument to useTokenBalance is a token, here we are passing a ref to the agent?

**Review by GabrielBuragev — APPROVED** [review]
Looks good! Very nice separation of the ethereum wallet & provider functionalities!

One tiny nitpick, other than that everything looks fine.

As discussed, i took over this PR and edited the last commit.
Now, the submit button will not be completely blocked when low liquidity was detected but instead a warning will be displayed to the user. Please note that this warning will only show up once the complete form is valid and filled.

**Comment by GabrielBuragev** [comment]
In order to provide a comprehensive solution for displaying the "time to expire" text i used the `date-fns` library.
It doesn't add much overhead (13.82kb gzipped) as it is tree-shaked on build-time. Here is the result of the build analyze:
Screenshot 2023-08-23 at 11 26 28

**Comment by GabrielBuragev** [comment]
@DBecane @andfletcher Here is an overview of the displayed text:
Screenshot 2023-08-23 at 09 54 24

Please let me know if this is fine by giving a thumbs up or otherwise leave your suggestions.

File: frontend/src/components/RequestDialog.vue [hunk 2]
```diff
@@ -199,7 +198,7 @@ function resetFormValidation() {
 }
 
 watch(signerAddress, (currSignerAddress, prevSignerAddress) => {
-  // Contract wallets should not prefill the target address as they need to be deployed separately per chain
+  // Contract providers should not prefill the target address as they need to be deployed separately per chain
```
[the remark this answers is not in the document]
Comment (GabrielBuragev) [reply frontend/src/components/RequestDialog.vue]: I changed this back to the old version.

**Pull request merged**